File: examples/python-examples/MeshOffsetWeighted.dox.py
from meshlib import mrmeshpy

# Create some mesh
mesh = mrmeshpy.makeTorus(primaryRadius=1.0,
                          secondaryRadius=0.1,
                          primaryResolution=16,
                          secondaryResolution=16)
verts_num = mesh.points.size()

# Create VertScalars obj with weights for every vertex
scalars = mrmeshpy.VertScalars(verts_num)
for i in range(verts_num):
    weight = abs(mesh.points.vec[i].x / 5) # Individual extra offset sizes for points
    scalars.vec[i] = weight

# ===params

params = mrmeshpy.WeightedShell.ParametersMetric()
# Algorithm is voxel based, voxel size affects performance and form of result mesh
params.voxelSize = mrmeshpy.suggestVoxelSize(mesh, 10000)
# common basic offset applied for all point
# Vertex-specific weighted offsets applied after the basic one
params.offset = 0.2
params.dist.maxWeight = max(scalars.vec) # should always have maximum between weights provided

# ===offset
res = mrmeshpy.WeightedShell.meshShell(mesh, scalars, params)

mrmeshpy.saveMesh(res, "offset_weighted.ctm")

File: examples/cpp-examples/MeshOffsetWeighted.dox.cpp
#include "MRMesh/MRTorus.h"
#include "MRMesh/MRMesh.h"
#include "MRMesh/MRParallelFor.h"
#include "MRVoxels/MRWeightedPointsShell.h"
#include "MRVoxels/MROffset.h"
#include "MRMesh/MRMeshSave.h"
#include <iostream>

int main()
{
    // Create some mesh
    auto mesh = MR::makeTorus();

    // Create VertScalars obj with weights for every vertex
    auto vertSize = mesh.topology.vertSize();
    MR::VertScalars scalars( vertSize );
    MR::ParallelFor( scalars, [&] ( MR::VertId v )
    {
        scalars[v] = std::abs( mesh.points[v].x / 5.0f ); // Individual extra offset sizes for points
    } );

    auto params = MR::WeightedShell::ParametersMetric();
    // Algorithm is voxel based, voxel size affects performance and form of result mesh
    params.voxelSize = MR::suggestVoxelSize( mesh, 1000 );
    // common basic offset applied for all point
    // Vertex-specific weighted offsets applied after the basic one
    params.offset = 0.2f;
    params.dist.maxWeight = *std::max_element( MR::begin( scalars ), MR::end( scalars ) ); // should always have maximum between weights provided
 
    auto res = MR::WeightedShell::meshShell( mesh, scalars, params );
    if ( !res.has_value() )
    {
        std::cerr << res.error();
        return 1;
    }
 
    auto saveRes = MR::MeshSave::toAnySupportedFormat( *res, "offset_weighted.ctm" );
    if ( !saveRes.has_value() )
    {
        std::cerr << saveRes.error();
        return 1;
    }
    return 0;
}
